src/org/traccar/protocol/NavisProtocolDecoder.java

```diff
@@ -272,70 +316,426 @@ private Object processArray(DeviceSession deviceSession, Channel channel, ByteBu
         return positions;
     }
 
+    private void skipFlexField(int index, ByteBuf buf) {
+        if (index < FLEX_FIELDS_SIZES.length) {
+            buf.skipBytes(FLEX_FIELDS_SIZES[index]);
+        }
+    }
+
+    private ParseResult parseFlexPosition(DeviceSession deviceSession, ByteBuf buf) {
+
+        Position position = new Position(getProtocolName());
+
+        position.setDeviceId(deviceSession.getDeviceId());
+
+        long index = 0;
+        int status = 0;
+        short input = 0;
+        short output = 0;
+
+        for (int i = 0; i < flexBitfieldDataSize; i++) {
+            if ((flexBitfield[(int) (i / 8)] & (0x80 >> i % 8)) == 0) {
```

Comment:
What does this mean?

Reply:
Here we check the specified bit in a bit field of a large-length stored in an byte array. I did a little refactoring to make the code clearer and at the same time I removed code duplication.

Reply:
I still don't fully understand it. Can you please provide more details and ideally some examples, or point me to the documentation.

Reply:
Protocol description is available on the manufacturer's website: https://navtelecom.ru/images/dokumentacia/protocol_of_information_exchange_navtelecom_v5.5_(160817).pdf
We are interested in the pages 15-16. This suggests that the filling of each bit of the bit field "bitfield" begins from the 7th bit. As a result, we must traversing every byte in the bit field starting from the 0th and every bit in the byte starting from the 7th, if we assume that the low bit is on the right.